#### Practise
##### Date: 2025-12-19


| Позначення     | Пояснення                             |
| -------------- | ------------------------------------- |
| n_A, n_B       | кількість користувачів у групах A і B |
| x_A, x_B       | кількість успіхів (купили, клікнули)  |
| p_A, p_B       | конверсія в групах                    |
| mean_A, mean_B | середнє значення метрики              |
| std_A, std_B   | стандартне відхилення                 |
| delta          | різниця між B і A                     |
| alpha          | рівень значущості (зазвичай 0.05)     |

---

#### z-test (для КОНВЕРСІЙ, 0/1)

```
p = x / n
```

---

#### Різниця конверсій

```
delta = p_B - p_A
```

---

#### Об’єднана (pooled) конверсія

```
p_pooled = (x_A + x_B) / (n_A + n_B)
```

---

#### Стандартна похибка (SE)

```
SE = sqrt( p_pooled * (1 - p_pooled) * (1/n_A + 1/n_B) )
```

---

#### z-статистика

```
z = delta / SE
```

---

#### p-value (двосторонній тест)

```
p_value = 2 * (1 - NormalCDF( |z| ))
```

---

### 95% довірчий інтервал для різниці конверсій
```
SE_diff = sqrt(
    p_A * (1 - p_A) / n_A +
    p_B * (1 - p_B) / n_B
)

CI = delta ± 1.96 * SE_diff
```

---

#### t-test (для ЧИСЛОВИХ МЕТРИК)

#### Різниця середніх

```
delta = mean_B - mean_A
```

---

#### Стандартна похибка (Welch t-test)

```
SE = sqrt( (std_A^2 / n_A) + (std_B^2 / n_B) )
```

---

#### t-статистика

```
t = delta / SE
```

---

### 95% довірчий інтервал для різниці середніх

```
CI = delta ± t_critical * SE
```

(де t_critical ≈ 2 при великих вибірках)

---



| Тип метрики       | Приклад          | Тест           |
| ----------------- | ---------------- | -------------- |
| Бінарна (0/1)     | купив / не купив | z-test         |
| Числова           | час, гроші       | t-test (Welch) |
| Дуже малі вибірки | <5 успіхів       | Fisher exact   |
| Сильний перекіс   | ARPU             | bootstrap      |

---

#### Інтерпретація z та t

| Значення | Що означає    |
| -------- | ------------- |
| ≈ 0      | різниці немає |
| ≈ 1      | слабкий ефект |
| ≈ 2      | на межі       |
| ≥ 2      | значущий      |
| ≥ 3      | сильний ефект |

---

#### Інтерпретація p-value

| p-value  | Значення            |
| -------- | ------------------- |
| < 0.01   | дуже сильний доказ  |
| < 0.05   | статистично значуще |
| 0.05–0.1 | сумнівно            |
| ≥ 0.1    | доказів немає       |

❗ p-value **НЕ показує розмір ефекту**

---

#### Інтерпретація довірчого інтервалу

| CI        | Висновок             |
| --------- | -------------------- |
| весь > 0  | ефект позитивний     |
| весь < 0  | ефект негативний     |
| містить 0 | результат незначущий |
| вузький   | оцінка точна         |
| широкий   | мало даних / шум     |

---

#### Гіпотези (шаблон)

##### Двосторонній тест

```
H0: A = B
H1: A ≠ B
```

##### Односторонній тест

```
H0: B ≤ A
H1: B > A
```

---

### z-test вручну

```python
z = delta / SE
p_value = 2 * (1 - stats.norm.cdf(abs(z)))
```

---

### t-test (Welch)

```python
from scipy import stats
stats.ttest_ind(A, B, equal_var=False)
```

---

| Ситуація         | Що робити                         |
| ---------------- | --------------------------------- |
| p-value < 0.05   | ефект не випадковий               |
| CI містить 0     | рішення ризиковане                |
| ефект малий      | дивимось на гроші                 |
| незначущий тест  | або мало даних, або слабкий ефект |
| велика дисперсія | t-test може бути ненадійним       |



In [1]:
# Task 
# Порахувати середнє у групі A і B
# Застосувати Welch t-test
# Подивитися на p-value
# Зробити висновок

import numpy as np
from scipy import stats

# Дані
A = np.array([10, 12, 11, 9, 10])
B = np.array([13, 14, 12, 15, 14])

# Крок 1: середні значення
mean_A = np.mean(A) # use mean
mean_B = np.mean(B) # use mean

print("Середній чек A:", mean_A)
print("Середній чек B:", mean_B)

# Крок 2: t-test
t_stat, p_value = stats.ttest_ind(A, B, equal_var=False)
print("t-статистика:", t_stat)
print("p-значення:", p_value)



Середній чек A: 10.4
Середній чек B: 13.6
t-статистика: -4.437601569801831
p-значення: 0.00217487819160667


📌 Разница между A и B статистически значима
📌 Вариант B лучше, потому что:
средний чек выше
вероятность, что это случайность — ~0.2%

5. Простыми словами
Мы почти наверняка можем сказать, что вариант B действительно увеличивает средний чек, а не просто «повезло с выборкой».

In [4]:
# Task 

# Кроки
# 1. Порахувати середній час
# 2. Застосувати Welch t-test
# 3. Оцінити p-value
# 4. Інтерпретувати результат

A = np.array([40, 42, 38, 41, 39, 40])
B = np.array([43, 44, 45, 41, 42, 43])
print("Середнє A:", mean_A)
print("Середнє B:", mean_B)
# Крок 1. Порахувати mean, use mean
mean_A = np.mean(A) # use mean
mean_B = np.mean(B)


# Крок 2. Зробити тест, use stats.ttest_ind(equal_var=False)

t_stat, p_value = stats.ttest_ind(A, B, equal_var=False)

print("t_stat =", t_stat)
print("p_value =", p_value)


Середнє A: 40.0
Середнє B: 43.0
t_stat = -3.6742346141747673
p_value = 0.004287014191860836


1. p-значение 0.00429 < 0.05
даже < 0.01
👉 Разница статистически значима с высокой уверенностью
(вероятность случайности ≈ 0.4%)

2. t-статистика
Минус ⇒ первый вариант хуже, чем второй (A < B)
|t| = 3.67 ⇒ эффект заметный, не шум

Вывод
📌 Нулевая гипотеза отвергается
📌 Вариант B статистически значимо лучше варианта A

Коротко «по-человечески»
Мы уверены, что разница между вариантами реальная, а не случайная. Вариант B выигрывает.

In [5]:
# Task 

# 1. Порахувати конверсію
# 2. Застосувати z-test для пропорцій
# 3. Подивитися на p-value
# 4. Зробити висновок

import math
from scipy.stats import norm

nA = 100
xA = 10

nB = 100
xB = 18

# Крок 1. Порахувати конверсію 
pA = xA / nA # you code 
pB = xB / nB # you code 
print("Конверсія A:", pA)
print("Конверсія B:", pB)

# Крок 2: z-test
p_pool = (xA + xB) / (nA + nB)  # you code
SE = math.sqrt(p_pool * (1 - p_pool) * (1/nA + 1/nB))
z = (pA - pB) / SE  # you code
p_value = 2 * (1 - norm.cdf(abs(z)))
print("z-статистика:", z)
print("p-значення:", p_value)


Конверсія A: 0.1
Конверсія B: 0.18
z-статистика: -1.6302782918784444
p-значення: 0.10304269342469352


1. Конверсии A = 0.10 (10%) B = 0.18 (18%)
👉 Вариант B выглядит лучше по конверсии на глаз.

2. z-статистика = −1.63
Минус ⇒ конверсия A ниже, чем B
По модулю значение не очень большое → сигнал слабый

3. p-значение = 0.103
Стандартный порог: 0.05
0.103 > 0.05
👉 Разница НЕ статистически значима

4. Итоговый вывод
📌 Нельзя отвергнуть нулевую гипотезу
📌 Мы не можем уверенно сказать, что B действительно лучше A по конверсии
📌 Наблюдаемая разница может быть случайной

5. Простыми словами
Да, у B конверсия выше (18% vs 10%), но данных недостаточно, чтобы быть уверенными, что это не просто флуктуация.

In [ ]:
# Task 1. Ви тестуєте новий текст на кнопці оплати.
# Метрика: purchase conversion (купив / не купив) — бінарна.

# Дані за час тесту:
# A (контроль): 8 000 користувачів, 720 оплат
# B (тест): 8 000 користувачів, 784 оплати

# Крок 0. Формулюємо гіпотези

#  різниці немає
#  різниця є

# Двосторонній тест, бо зміна може як покращити, так і погіршити результат.


In [ ]:
# Крок 1. Рахуємо конверсії та дельта(pB​−pA)
# Крок 2. Об’єднана конверсія (p = (pa + pb) / count A + count B)
# Крок 3. Стандартна похибка(SE)
# Крок 4. z-статистика дельта / SE
# Крок 5. p-value (двосторонній)

In [7]:

import numpy as np
from scipy import stats

nA, xA = 8000, 720
nB, xB = 8000, 784

pA = xA / nA
pB = xB / nB
print("Конверсія A:", pA)
print("Конверсія B:", pB)

delta = pB - pA

p_pool = (xA + xB) / (nA + nB)
se = np.sqrt(p_pool * (1 - p_pool) * (1/nA + 1/nB))

z = delta / se
p_value = 2 * (1 - stats.norm.cdf(abs(z)))

print("z-статистика:", z)
print("p-значення:", p_value)
# Висновок: якщо p-value < 0.05, відхиляємо H0, інакше не відхиляємо H0.

Конверсія A: 0.09
Конверсія B: 0.098
z-статистика: 1.7337718886061455
p-значення: 0.08295856537338397


1. Разница в конверсии
Абсолютная разница: +0.8 п.п.
Относительный рост: ~+8.9%
👉 B немного лучше A, но эффект маленький.

2. p-значение
0.083 > 0.05
👉 При стандартном уровне значимости α = 0.05
результат НЕ статистически значим.
⚠️ Но:
0.083 < 0.10
👉 При уровне значимости α = 0.10 результат был бы гранично значимым.

3. z-статистика
Положительная ⇒ B > A
|z| = 1.73 — слабый/средний сигнал, не дотягивает до уверенного вывода

Итог

📌 При α = 0.05: доказательств недостаточно
📌 Есть тенденция в пользу B, но уверенности нет
📌 Вероятно, не хватает выборки

Простыми словами

B выглядит чуть лучше, но мы пока не можем с уверенностью сказать, что он действительно выигрывает. Нужно больше данных.

Что можно сделать

📈 Увеличить размер выборки

🧮 Посчитать доверительный интервал

🎯 Проверить, насколько такая разница вообще бизнес-значима

--- 
Task 2. Ви тестуєте новий порядок тарифів.

Метрика: оплата.

Дані:
- A: 12 000 користувачів, 1 020 оплат
- B: 12 000 користувачів, 1 128 оплат

Проведіть z-test (двосторонній, 𝛼=0.05 α=0.05)
та зробіть висновок. Додатково порахуйте CI

In [ ]:
# Крок 1. Конверсії
# Крок 2. Pooled conversion
# Крок 3. Стандартна похибка
# Крок 4. z-статистика
# Крок 5. p-value



In [8]:
import numpy as np
from scipy import stats

nA, xA = 12000, 1200
nB, xB = 12000, 1128

pA = xA / nA
pB = xB / nB
print("Конверсія A:", pA)
print("Конверсія B:", pB)

delta = pB - pA

p_pool = (xA + xB) / (nA + nB)
se = np.sqrt(p_pool * (1 - p_pool) * (1/nA + 1/nB))

z = delta / se
p_value = 2 * (1 - stats.norm.cdf(abs(z)))

print("z-статистика:", z)
print("p-значення:", p_value)

Конверсія A: 0.1
Конверсія B: 0.094
z-статистика: -1.570352438632377
p-значення: 0.11633314085389612


Конверсія A: 0.10 (10%)
Конверсія B: 0.094 (9.4%)

z-статистика: −1.57
p-значение: 0.116

Что это значит
1. Разница

Абсолютная разница: −0.6 п.п.
B немного хуже, чем A

2. p-значение
0.116 > 0.05
даже > 0.10
👉 Разница НЕ статистически значима

3. z-статистика
Отрицательная ⇒ B < A
|z| = 1.57 — слабый сигнал

Итоговый вывод

📌 Нулевая гипотеза не отвергается
📌 Мы не можем утверждать, что A лучше B
📌 Наблюдаемая разница может быть случайной

Простыми словами
Хотя у A конверсия чуть выше, данных недостаточно, чтобы уверенно сказать, что B действительно хуже.

Рекомендации

📈 Собрать больше данных

🎯 Оценить, важна ли разница в 0.6 п.п. для бизнеса

📊 Посмотреть доверительный интервал (скорее всего, он включает 0)